Index: target/linux/ramips/image/Makefile
===================================================================
--- target/linux/ramips/image/Makefile	(revision 43770)
+++ target/linux/ramips/image/Makefile	(working copy)
@@ -724,6 +724,7 @@
 na930_mtd_size=20971520
 Image/Build/Profile/NA930=$(call BuildFirmware/CustomFlash/$(1),$(1),na930,NA930,$(na930_mtd_size))
 Image/Build/Profile/MZK-750DHP=$(call BuildFirmware/Default8M/$(1),$(1),mzk-750dhp,MZK-750DHP)
+Image/Build/Profile/XIAOMI-MIWIFI-MINI=$(call BuildFirmware/Default16M/$(1),$(1),xiaomi-miwifi-mini,XIAOMI-MIWIFI-MINI)
 Image/Build/Profile/Y1=$(call BuildFirmware/Default16M/$(1),$(1),Lenovo-y1,Y1)
 Image/Build/Profile/Y1S=$(call BuildFirmware/Default16M/$(1),$(1),Lenovo-y1s,Y1S)
 
@@ -740,6 +741,7 @@
 	$(call Image/Build/Profile/WHR1166D,$(1))
 	$(call Image/Build/Profile/MZK-750DHP,$(1))
 	$(call Image/Build/Profile/NA930,$(1))
+	$(call Image/Build/Profile/XIAOMI-MIWIFI-MINI,$(1))
 	$(call Image/Build/Profile/Y1,$(1))
 	$(call Image/Build/Profile/Y1S,$(1))
 endef
Index: target/linux/ramips/dts/XIAOMI-MIWIFI-MINI.dts
===================================================================
--- target/linux/ramips/dts/XIAOMI-MIWIFI-MINI.dts	(revision 0)
+++ target/linux/ramips/dts/XIAOMI-MIWIFI-MINI.dts	(revision 0)
@@ -0,0 +1,154 @@
+/dts-v1/;
+
+/include/ "mt7620a.dtsi"
+
+/ {
+	compatible = "XIAOMI MIWIFI MINI", "ralink,mt7620a-soc";
+	model = "Xiaomi MiWiFi Mini";
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	palmbus@10000000 {
+		gpio0: gpio@600 {
+			status = "okay";
+		};
+
+		gpio1: gpio@638 {
+			status = "okay";
+		};
+
+		gpio2: gpio@660 {
+			status = "okay";
+		};
+
+		spi@b00 {
+			status = "okay";
+
+			m25p80@0 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+				compatible = "w25q128";
+				reg = <0 0>;
+				linux,modalias = "m25p80", "w25q128";
+				spi-max-frequency = <10000000>;
+
+				partition@0 {
+					label = "u-boot";
+					reg = <0x0 0x30000>;
+				};
+
+				partition@30000 {
+					label = "u-boot-env";
+					reg = <0x30000 0x10000>;
+					read-only;
+				};
+
+				factory: partition@40000 {
+					label = "factory";
+					reg = <0x40000 0x10000>;
+					read-only;
+				};
+
+				partition@50000 {
+					label = "firmware";
+					reg = <0x50000 0xf80000>;
+				};
+
+				partition@fd0000 {
+					label = "crash";
+					reg = <0xfd0000 0x10000>;
+				};
+
+				partition@fe0000 {
+					label = "reserved";
+					reg = <0xfe0000 0x10000>;
+					read-only;
+				};
+
+				partition@ff0000 {
+					label = "Bdata";
+					reg = <0xff0000 0x10000>;
+				};
+			};
+		};
+	};
+
+	ehci@101c0000 {
+		status = "okay";
+	};
+
+	ohci@101c1000 {
+		status = "okay";
+	};
+
+	ethernet@10100000 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&ephy_pins>;
+		mtd-mac-address = <&factory 0x4>;
+		ralink,port-map = "llllw";
+		#ralink,port-map = "lllwl";
+	};
+
+	wmac@10180000 {
+		ralink,mtd-eeprom = <&factory 0>;
+	};
+
+	pcie@10140000 {
+		status = "okay";
+
+		pcie-bridge {
+			mt76@0,0 {
+				reg = <0x0000 0 0 0 0>;
+				device_type = "pci";
+				mediatek,mtd-eeprom = <&factory 0x8000>;
+				mediatek,2ghz = <0>;
+			};
+		};
+	};
+
+	pinctrl {
+		state_default: pinctrl0 {
+			gpio {
+				ralink,group = "i2c", "rgmii1";
+				ralink,function = "gpio";
+			};
+			pa {
+				ralink,group = "pa";
+				ralink,function = "pa";
+			};
+		};
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		blue {
+			label = "xiaomi:blue:status";
+			gpios = <&gpio1 0 1>;
+		};
+
+		yellow {
+			label = "xiaomi:yellow:status";
+			gpios = <&gpio1 2 1>;
+		};
+
+		red {
+			label = "xiaomi:red:status";
+			gpios = <&gpio1 5 1>;
+		};
+	};
+
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
+		reset {
+			label = "reset";
+			gpios = <&gpio1 6 0>;
+			linux,code = <0x198>;
+		};
+	};
+};
Index: target/linux/ramips/mt7620a/profiles/xiaomi-miwifi-mini.mk
===================================================================
--- target/linux/ramips/mt7620a/profiles/xiaomi-miwifi-mini.mk	(revision 0)
+++ target/linux/ramips/mt7620a/profiles/xiaomi-miwifi-mini.mk	(revision 0)
@@ -0,0 +1,20 @@
+#
+# Copyright (C) 2015 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/XIAOMI-MIWIFI-MINI
+ NAME:=Xiaomi MiWiFi Mini
+ PACKAGES:=\
+	kmod-usb-core kmod-usb-dwc2 kmod-usb2 kmod-usb-ohci \
+	kmod-mt76 \
+	kmod-usb-storage block-mount mountd kmod-fs-ext4 \
+	ppp-mod-pppoe e2fsprogs kmod-ledtrig-usbdev fdisk
+endef
+
+define Profile/XIAOMI-MIWIFI-MINI/Description
+ Support for Xiaomi MiWiFi Mini routers
+endef
+$(eval $(call Profile,XIAOMI-MIWIFI-MINI))
Index: target/linux/ramips/base-files/lib/ramips.sh
===================================================================
--- target/linux/ramips/base-files/lib/ramips.sh	(revision 43770)
+++ target/linux/ramips/base-files/lib/ramips.sh	(working copy)
@@ -325,6 +325,9 @@
 	*"ZBT WR8305RT")
 		name="wr8305rt"
 		;;
+	*"Xiaomi MiWiFi Mini")
+		name="xiaomi-miwifi-mini"
+		;;
 	*"Lenovo Y1")
 		name="y1"
 		;;
Index: target/linux/ramips/base-files/lib/upgrade/platform.sh
===================================================================
--- target/linux/ramips/base-files/lib/upgrade/platform.sh	(revision 43770)
+++ target/linux/ramips/base-files/lib/upgrade/platform.sh	(working copy)
@@ -98,6 +98,7 @@
 	x5 |\
 	x8 |\
 	zbt-wa05 |\
+	xiaomi-miwifi-mini |\
 	y1 |\
 	y1s)
 		[ "$magic" != "27051956" ] && {
Index: target/linux/ramips/base-files/etc/uci-defaults/01_leds
===================================================================
--- target/linux/ramips/base-files/etc/uci-defaults/01_leds	(revision 43770)
+++ target/linux/ramips/base-files/etc/uci-defaults/01_leds	(working copy)
@@ -203,6 +203,9 @@
 		set_usb_led "wr8305rt:usb"
 		set_wifi_led "wr8305rt:wifi"
 		;;
+	xiaomi-miwifi-mini)
+		ucidef_set_led_default "power" "power" "xiaomi:red:status" "1" 
+		;;
 	y1 |\
 	y1s)
 		ucidef_set_led_default "power" "power" "lenovo:blue:power" "1"
Index: target/linux/ramips/base-files/etc/uci-defaults/02_network
===================================================================
--- target/linux/ramips/base-files/etc/uci-defaults/02_network	(revision 43770)
+++ target/linux/ramips/base-files/etc/uci-defaults/02_network	(working copy)
@@ -96,6 +96,7 @@
 		ucidef_add_switch_vlan "switch0" "2" "4 9t"
 		;;
 
+	xiaomi-miwifi-mini |\
 	3g-6200n | \
 	dir-610-a1 | \
 	dir-300-b7 | \
Index: target/linux/ramips/base-files/etc/diag.sh
===================================================================
--- target/linux/ramips/base-files/etc/diag.sh	(revision 43770)
+++ target/linux/ramips/base-files/etc/diag.sh	(working copy)
@@ -188,6 +188,9 @@
 	na930)
 		status_led="na930:blue:power"
 		;;
+	xiaomi-miwifi-mini)
+		status_led="xiaomi:red:status"
+		;;
 	y1 | \
 	y1s)
 		status_led="lenovo:blue:power"
Index: target/linux/ramips/patches-3.10/0402-mtd-gd25q.patch
===================================================================
--- target/linux/ramips/patches-3.10/0402-mtd-gd25q.patch	(revision 0)
+++ target/linux/ramips/patches-3.10/0402-mtd-gd25q.patch	(revision 0)
@@ -0,0 +1,10 @@
+--- a/drivers/mtd/devices/m25p80.c
++++ b/drivers/mtd/devices/m25p80.c
+@@ -984,6 +984,7 @@ static const struct spi_device_id m25p_ids[] = {
+ 	{ "w25q80bl", INFO(0xef4014, 0, 64 * 1024,  16, SECT_4K) },
+ 	{ "w25q128", INFO(0xef4018, 0, 64 * 1024, 256, SECT_4K) },
+ 	{ "w25q256", INFO(0xef4019, 0, 64 * 1024, 512, SECT_4K) },
++	{ "gd25q128b", INFO(0xc84018, 0, 64 * 1024, 256, SECT_4K)},
+ 
+ 	/* Catalyst / On Semiconductor -- non-JEDEC */
+ 	{ "cat25c11", CAT25_INFO(  16, 8, 16, 1) },
